overlord/fdestate/fdestate.go

```diff
@@ -238,40 +240,44 @@ func initializeState(st *state.State) error {
 		}
 	}
 
+	keyFile := device.DataSealedKeyUnder(dirs.SnapSaveDir)
+	runCounterHandle, err := secbootGetPCRHandle(devpData, "default", keyFile)
+	if err != nil {
+		return fmt.Errorf("cannot obtain counter handle: %w", err)
+	}
+
+	fallbackKeyFile := device.FallbackDataSealedKeyUnder(dirs.SnapSaveDir)
+	fallbackCounterHandle, err := secbootGetPCRHandle(devpData, "default-fallback", fallbackKeyFile)
+	if err != nil {
+		return fmt.Errorf("cannot obtain counter handle: %w", err)
+	}
+
+	fallbackSaveKeyFile := device.FallbackDataSealedKeyUnder(dirs.SnapSaveDir)
+	fallbackSaveCounterHandle, err := secbootGetPCRHandle(devpSave, "default-fallback", fallbackSaveKeyFile)
+	if err != nil {
+		return fmt.Errorf("cannot obtain counter handle: %w", err)
+	}
+
+	if fallbackCounterHandle != fallbackSaveCounterHandle {
+		return fmt.Errorf("the recover data and save keys do not use the same counter handle")
+	}
+
 	// Note that Parameters will be updated on first update
 	s.KeyslotRoles = map[string]KeyslotRoleInfo{
 		// TODO: use a constant
 		"run": {
-			PrimaryKeyID: 0,
-			// FIXME: from Chris: Rather than hardcoding
-			// an index value, I'd prefer us to adopt the
-			// approach of picking a random index from a
-			// small acceptable range. I could add an API
-			// in secboot for that, or we could do it
-			// here.
-			TPM2PCRPolicyRevocationCounter: secboot.RunObjectPCRPolicyCounterHandle,
+			PrimaryKeyID:                   0,
+			TPM2PCRPolicyRevocationCounter: runCounterHandle,
 		},
 		// TODO: use a constant
 		"run+recover": {
-			PrimaryKeyID: 0,
-			// FIXME: from Chris: Rather than hardcoding
-			// an index value, I'd prefer us to adopt the
-			// approach of picking a random index from a
-			// small acceptable range. I could add an API
-			// in secboot for that, or we could do it
-			// here.
-			TPM2PCRPolicyRevocationCounter: secboot.RunObjectPCRPolicyCounterHandle,
+			PrimaryKeyID:                   0,
+			TPM2PCRPolicyRevocationCounter: runCounterHandle,
 		},
 		// TODO: use a constant
 		"recover": {
-			PrimaryKeyID: 0,
-			// FIXME: from Chris: Rather than hardcoding
-			// an index value, I'd prefer us to adopt the
-			// approach of picking a random index from a
-			// small acceptable range. I could add an API
-			// in secboot for that, or we could do it
-			// here.
-			TPM2PCRPolicyRevocationCounter: secboot.FallbackObjectPCRPolicyCounterHandle,
+			PrimaryKeyID:                   0,
+			TPM2PCRPolicyRevocationCounter: fallbackCounterHandle,
```

Comment:
Is there any scenario where we expect this to be different to `runCounterHandle`?

Reply:
Yes, for legacy keys. We still initialize some state for them. Otherwise we cannot reseal older installations. Also there is the case where we do not use "plainkey", then we cannot really verify that the primary key matches, so in theory the handle might not match. Checking the handles separately, makes it just a bit cleaner for all corner cases. Though we could add some warning when we find they do not match.